File: sidequest/pipeline-core/annotators/semantic_annotator.py
"""
Semantic Annotator - Stage 4 of the Duplicate Detection Pipeline

Extracts rich semantic metadata from code blocks to enable Layer 3
(semantic similarity) grouping.

Semantic tags are organized into four categories:
- Operations: What the code does (filter, map, validate, fetch, etc.)
- Domains: What domain concepts are involved (user, auth, payment, etc.)
- Patterns: What code patterns are used (guard_clause, null_check, etc.)
- Data Types: What data types are processed (array, object, promise, etc.)
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..models.code_block import CodeBlock


@dataclass
class SemanticAnnotation:
    """Rich semantic metadata for a code block.

    Attributes:
        category: The semantic category (from CodeBlock)
        operations: Set of operations performed (filter, map, validate, etc.)
        domains: Set of domain concepts (user, auth, payment, etc.)
        patterns: Set of code patterns (guard_clause, null_check, etc.)
        data_types: Set of data types processed (array, object, promise, etc.)
        intent: High-level description of what the code does
    """

    category: str
    operations: set[str] = field(default_factory=set)
    domains: set[str] = field(default_factory=set)
    patterns: set[str] = field(default_factory=set)
    data_types: set[str] = field(default_factory=set)
    intent: str = ""

    def all_tags(self) -> set[str]:
        """Return all semantic tags as a single set."""
        return self.operations | self.domains | self.patterns | self.data_types

    def to_dict(self) -> dict:
        """Convert to dictionary for serialization."""
        return {
            'category': self.category,
            'operations': sorted(self.operations),
            'domains': sorted(self.domains),
            'patterns': sorted(self.patterns),
            'data_types': sorted(self.data_types),
            'intent': self.intent,
        }


# ---------------------------------------------------------------------------
# Pattern Libraries
# ---------------------------------------------------------------------------

# Array/collection operations
ARRAY_OPERATION_PATTERNS: dict[str, str] = {
    r'\.filter\s*\(': 'filter',
    r'\.map\s*\(': 'map',
    r'\.reduce\s*\(': 'reduce',
    r'\.find\s*\(': 'find',
    r'\.findIndex\s*\(': 'find',
    r'\.some\s*\(': 'some',
    r'\.every\s*\(': 'every',
    r'\.sort\s*\(': 'sort',
    r'\.includes\s*\(': 'includes',
    r'\.indexOf\s*\(': 'find',
    r'\.forEach\s*\(': 'iterate',
    r'for\s*\(\s*(?:const|let|var)\s+\w+\s+(?:of|in)': 'iterate',
    r'for\s*\(\s*(?:let|var)\s+\w+\s*=': 'iterate',
    r'while\s*\(': 'iterate',
    r'\.flat\s*\(': 'flatten',
    r'\.flatMap\s*\(': 'flatten',
    r'\.concat\s*\(': 'concat',
    r'\.slice\s*\(': 'slice',
    r'\.splice\s*\(': 'splice',
    r'\.push\s*\(': 'append',
    r'\.pop\s*\(': 'remove',
    r'\.shift\s*\(': 'remove',
    r'\.unshift\s*\(': 'prepend',
}

# CRUD/data operations
CRUD_OPERATION_PATTERNS: dict[str, str] = {
    r'\.(get|fetch|read|load|retrieve)\s*\(': 'read',
    r'\.(post|create|insert|add|save|write)\s*\(': 'create',
    r'\.(put|update|patch|modify|set)\s*\(': 'update',
    r'\.(delete|remove|destroy|clear)\s*\(': 'delete',
    r'fetch\s*\(': 'fetch',
    r'axios\.(get|post|put|patch|delete)\s*\(': 'fetch',
    r'http\.(get|post|put|patch|delete)\s*\(': 'fetch',
}

# Transformation operations
TRANSFORM_OPERATION_PATTERNS: dict[str, str] = {
    r'JSON\.parse\s*\(': 'parse',
    r'JSON\.stringify\s*\(': 'serialize',
    r'\.toString\s*\(': 'transform',
    r'\.toUpperCase\s*\(': 'transform',
    r'\.toLowerCase\s*\(': 'transform',
    r'\.trim\s*\(': 'transform',
    r'\.split\s*\(': 'split',
    r'\.join\s*\(': 'join',
    r'\.replace\s*\(': 'replace',
    r'parseInt\s*\(': 'parse',
    r'parseFloat\s*\(': 'parse',
    r'Number\s*\(': 'transform',
    r'String\s*\(': 'transform',
    r'Boolean\s*\(': 'transform',
    r'Object\.keys\s*\(': 'extract',
    r'Object\.values\s*\(': 'extract',
    r'Object\.entries\s*\(': 'extract',
    r'Object\.assign\s*\(': 'merge',
    r'\.\.\.\w+': 'spread',
}

# Validation operations
VALIDATION_OPERATION_PATTERNS: dict[str, str] = {
    r'(validate|isValid|check|verify|assert)\s*\(': 'validate',
    r'\.test\s*\(': 'validate',
    r'\.match\s*\(': 'validate',
    r'schema\.(validate|parse|safeParse)\s*\(': 'validate',
    r'z\.\w+\s*\(': 'validate',
    r'joi\.\w+': 'validate',
    r'yup\.\w+': 'validate',
}

# Domain patterns
DOMAIN_PATTERNS: dict[str, str] = {
    r'\b(user|users|account|accounts|profile|profiles|member)\b': 'user',
    r'\b(auth|authentication|login|logout|signin|signout|token|session|jwt|oauth)\b': 'auth',
    r'\b(password|credential|secret|apikey|api_key)\b': 'auth',
    r'\b(payment|charge|invoice|billing|subscription|stripe|paypal)\b': 'payment',
    r'\b(order|orders|cart|checkout|purchase)\b': 'commerce',
    r'\b(email|mail|notification|alert|notify|message|sms)\b': 'notification',
    r'\b(file|files|upload|download|attachment|blob|storage)\b': 'file',
    r'\b(database|db|query|record|table|collection|document)\b': 'database',
    r'\b(prisma|mongoose|sequelize|typeorm|knex)\b': 'database',
    r'\b(cache|redis|memcached|cached)\b': 'cache',
    r'\b(queue|job|jobs|worker|task|tasks|bull|rabbitmq)\b': 'queue',
    r'\b(api|endpoint|route|routes|request|response|req|res)\b': 'api',
    r'\b(webhook|webhooks|callback|hook)\b': 'webhook',
    r'\b(event|events|emit|publish|subscribe|listener)\b': 'event',
    r'\b(log|logs|logger|logging|trace|debug|info|warn|error)\b': 'logging',
    r'\b(config|configuration|settings|options|env|environment)\b': 'config',
    r'\b(test|tests|spec|describe|it\s*\(|expect\s*\()\b': 'test',
}

# Code patterns
CODE_PATTERN_PATTERNS: dict[str, str] = {
    # Guard clause / early return
    r'if\s*\([^)]+\)\s*(?:return|throw)': 'guard_clause',
    r'if\s*\(\s*!\s*\w+\s*\)\s*(?:return|throw)': 'guard_clause',

    # Null/undefined checks
    r'===?\s*null\b': 'null_check',
    r'!==?\s*null\b': 'null_check',
    r'===?\s*undefined\b': 'null_check',
    r'!==?\s*undefined\b': 'null_check',
    r'\?\?': 'null_check',
    r'\?\s*\.': 'null_check',
    r'typeof\s+\w+\s*[!=]==?\s*["\']undefined["\']': 'null_check',

    # Error handling
    r'try\s*\{': 'error_handling',
    r'catch\s*\(': 'error_handling',
    r'\.catch\s*\(': 'error_handling',
    r'finally\s*\{': 'error_handling',
    r'throw\s+new\s+\w*Error': 'error_handling',

    # Retry logic
    r'retry|retries|attempts|maxAttempts|backoff': 'retry_logic',

    # Timeout handling
    r'timeout|setTimeout|setInterval|clearTimeout|clearInterval': 'timeout',

    # Async patterns
    r'async\s+': 'async_await',
    r'await\s+': 'async_await',
    r'\.then\s*\(': 'promise_chain',
    r'Promise\.(all|race|allSettled|any)\s*\(': 'promise_composition',
    r'new\s+Promise\s*\(': 'promise_creation',

    # Caching patterns
    r'cache\.(get|set|has|delete)': 'caching',
    r'memoize|memo|cached': 'caching',

    # Pagination
    r'\b(page|pages|offset|limit|cursor|skip|take)\b': 'pagination',

    # Batching
    r'\b(batch|batches|chunk|chunks)\b': 'batching',

    # Streaming
    r'\b(stream|streams|pipe|readable|writable)\b': 'streaming',

    # Locking/mutex
    r'\b(lock|unlock|mutex|semaphore)\b': 'locking',

    # Rate limiting
    r'\b(rateLimit|throttle|debounce)\b': 'rate_limiting',
}

# Data type patterns
DATA_TYPE_PATTERNS: dict[str, str] = {
    # Array
    r'\[\s*\]': 'array',
    r'\bArray\b': 'array',
    r'\.length\b': 'array',
    r'Array\.isArray\s*\(': 'array',
    r'\.push\s*\(': 'array',

    # Object
    r'\{\s*\}': 'object',
    r'\bObject\b': 'object',
    r'\.keys\s*\(': 'object',
    r'\.values\s*\(': 'object',
    r'\.entries\s*\(': 'object',
    r'\.hasOwnProperty\s*\(': 'object',

    # String
    r'["\'][^"\']*["\']': 'string',
    r'`[^`]*`': 'string',
    r'\.toString\s*\(': 'string',
    r'\.trim\s*\(': 'string',
    r'\.substring\s*\(': 'string',
    r'\.substr\s*\(': 'string',

    # Number
    r'\b\d+\.?\d*\b': 'number',
    r'Number\s*\(': 'number',
    r'parseInt\s*\(': 'number',
    r'parseFloat\s*\(': 'number',
    r'Math\.\w+': 'number',

    # Boolean
    r'\b(true|false)\b': 'boolean',
    r'Boolean\s*\(': 'boolean',

    # Date
    r'new\s+Date\s*\(': 'date',
    r'Date\.(now|parse)\s*\(': 'date',
    r'\.toISOString\s*\(': 'date',
    r'moment\s*\(': 'date',
    r'dayjs\s*\(': 'date',

    # Promise/async
    r'\bPromise\b': 'promise',
    r'\.then\s*\(': 'promise',
    r'async\s+': 'promise',
    r'await\s+': 'promise',

    # Null/undefined
    r'\bnull\b': 'null',
    r'\bundefined\b': 'undefined',

    # Map/Set
    r'new\s+Map\s*\(': 'map',
    r'new\s+Set\s*\(': 'set',
    r'\.has\s*\(': 'collection',

    # Buffer/Binary
    r'\bBuffer\b': 'buffer',
    r'ArrayBuffer': 'buffer',
    r'Uint8Array': 'buffer',

    # Regex
    r'/[^/]+/[gim]*': 'regex',
    r'new\s+RegExp\s*\(': 'regex',
}


class SemanticAnnotator:
    """Stage 4: Full semantic annotation of code blocks.

    Analyzes code blocks and extracts rich semantic metadata for use in
    Layer 3 (semantic similarity) grouping.

    Usage:
        annotator = SemanticAnnotator()
        annotation = annotator.annotate(code_block)

        # Access semantic tags
        print(annotation.operations)  # {'filter', 'map'}
        print(annotation.domains)     # {'user'}
        print(annotation.intent)      # 'filter+map|on:user'
    """

    def annotate(self, block: CodeBlock) -> SemanticAnnotation:
        """Analyze code block and extract semantic metadata.

        Uses pattern matching to identify:
        - Operations (filter, map, validate, etc.)
        - Domain concepts (user, auth, payment, etc.)
        - Code patterns (guard clause, error handling, etc.)
        - Data types being processed

        Args:
            block: CodeBlock to annotate

        Returns:
            SemanticAnnotation with extracted metadata
        """
        code = block.source_code
        tags = block.tags if hasattr(block, 'tags') else []
        category = block.category if hasattr(block, 'category') else 'unknown'

        # Handle enum values
        if hasattr(category, 'value'):
            category = category.value

        operations = self._extract_operations(code)
        domains = self._extract_domains(code, tags)
        patterns = self._extract_patterns(code)
        data_types = self._extract_data_types(code)
        intent = self._infer_intent(operations, domains, patterns)

        return SemanticAnnotation(
            category=category,
            operations=operations,
            domains=domains,
            patterns=patterns,
            data_types=data_types,
            intent=intent,
        )

    def _extract_operations(self, code: str) -> set[str]:
        """Extract operation types from code.

        Detects array operations, CRUD operations, transformations,
        and validation operations.
        """
        operations: set[str] = set()

        all_patterns = {
            **ARRAY_OPERATION_PATTERNS,
            **CRUD_OPERATION_PATTERNS,
            **TRANSFORM_OPERATION_PATTERNS,
            **VALIDATION_OPERATION_PATTERNS,
        }

        for pattern, op in all_patterns.items():
            if re.search(pattern, code, re.IGNORECASE):
                operations.add(op)

        return operations

    def _extract_domains(self, code: str, tags: list[str]) -> set[str]:
        """Extract domain concepts from code and tags.

        Analyzes both source code and existing tags for domain indicators.
        """
        domains: set[str] = set()

        # Combine code and tags for analysis
        text = code + ' ' + ' '.join(tags)

        for pattern, domain in DOMAIN_PATTERNS.items():
            if re.search(pattern, text, re.IGNORECASE):
                domains.add(domain)

        return domains

    def _extract_patterns(self, code: str) -> set[str]:
        """Identify code patterns.

        Detects common patterns like guard clauses, error handling,
        null checks, async patterns, etc.
        """
        patterns: set[str] = set()

        for pattern, name in CODE_PATTERN_PATTERNS.items():
            if re.search(pattern, code, re.IGNORECASE):
                patterns.add(name)

        return patterns

    def _extract_data_types(self, code: str) -> set[str]:
        """Identify data types being processed.

        Detects arrays, objects, strings, numbers, promises, etc.
        """
        data_types: set[str] = set()

        for pattern, dtype in DATA_TYPE_PATTERNS.items():
            if re.search(pattern, code):
                data_types.add(dtype)

        return data_types

    def _infer_intent(
        self,
        operations: set[str],
        domains: set[str],
        patterns: set[str],
    ) -> str:
        """Generate high-level intent description.

        Creates a structured intent string that summarizes what the code does.
        Format: operations|on:domains|with:patterns

        Examples:
            - "filter+map|on:user"
            - "validate|on:auth|with:guard_clause"
            - "fetch+parse|on:api"
        """
        parts: list[str] = []

        if operations:
            parts.append('+'.join(sorted(operations)))

        if domains:
            parts.append('on:' + '+'.join(sorted(domains)))

        if patterns:
            parts.append('with:' + '+'.join(sorted(patterns)))

        return '|'.join(parts) if parts else 'unknown'

File: sidequest/pipeline-core/annotators/semantic-annotator.ts
/**
 * Semantic Annotator - Fourth stage of the Duplicate Detection Pipeline
 *
 * Extracts rich semantic metadata from code blocks to enable
 * semantic similarity layer grouping.
 *
 * Ports Python semantic_annotator.py.
 */

import type { CodeBlock, SemanticAnnotation } from '../models/types.ts';

// ---------------------------------------------------------------------------
// Pattern Libraries
// ---------------------------------------------------------------------------
// All regex patterns use bounded quantifiers (\s{0,20} instead of \s*)
// to prevent ReDoS attacks.

const ARRAY_OPERATION_PATTERNS: Array<[RegExp, string]> = [
  [/\.filter\s{0,20}\(/i, 'filter'],
  [/\.map\s{0,20}\(/i, 'map'],
  [/\.reduce\s{0,20}\(/i, 'reduce'],
  [/\.find\s{0,20}\(/i, 'find'],
  [/\.findIndex\s{0,20}\(/i, 'find'],
  [/\.some\s{0,20}\(/i, 'some'],
  [/\.every\s{0,20}\(/i, 'every'],
  [/\.sort\s{0,20}\(/i, 'sort'],
  [/\.includes\s{0,20}\(/i, 'includes'],
  [/\.indexOf\s{0,20}\(/i, 'find'],
  [/\.forEach\s{0,20}\(/i, 'iterate'],
  [/for\s{0,20}\(\s{0,20}(?:const|let|var)\s{1,20}\w+\s{1,20}(?:of|in)/i, 'iterate'],
  [/for\s{0,20}\(\s{0,20}(?:let|var)\s{1,20}\w+\s{0,20}=/i, 'iterate'],
  [/while\s{0,20}\(/i, 'iterate'],
  [/\.flat\s{0,20}\(/i, 'flatten'],
  [/\.flatMap\s{0,20}\(/i, 'flatten'],
  [/\.concat\s{0,20}\(/i, 'concat'],
  [/\.slice\s{0,20}\(/i, 'slice'],
  [/\.splice\s{0,20}\(/i, 'splice'],
  [/\.push\s{0,20}\(/i, 'append'],
  [/\.pop\s{0,20}\(/i, 'remove'],
  [/\.shift\s{0,20}\(/i, 'remove'],
  [/\.unshift\s{0,20}\(/i, 'prepend'],
];

const CRUD_OPERATION_PATTERNS: Array<[RegExp, string]> = [
  [/\.(get|fetch|read|load|retrieve)\s{0,20}\(/i, 'read'],
  [/\.(post|create|insert|add|save|write)\s{0,20}\(/i, 'create'],
  [/\.(put|update|patch|modify|set)\s{0,20}\(/i, 'update'],
  [/\.(delete|remove|destroy|clear)\s{0,20}\(/i, 'delete'],
  [/fetch\s{0,20}\(/i, 'fetch'],
  [/axios\.(get|post|put|patch|delete)\s{0,20}\(/i, 'fetch'],
  [/http\.(get|post|put|patch|delete)\s{0,20}\(/i, 'fetch'],
];

const TRANSFORM_OPERATION_PATTERNS: Array<[RegExp, string]> = [
  [/JSON\.parse\s{0,20}\(/i, 'parse'],
  [/JSON\.stringify\s{0,20}\(/i, 'serialize'],
  [/\.toString\s{0,20}\(/i, 'transform'],
  [/\.toUpperCase\s{0,20}\(/i, 'transform'],
  [/\.toLowerCase\s{0,20}\(/i, 'transform'],
  [/\.trim\s{0,20}\(/i, 'transform'],
  [/\.split\s{0,20}\(/i, 'split'],
  [/\.join\s{0,20}\(/i, 'join'],
  [/\.replace\s{0,20}\(/i, 'replace'],
  [/parseInt\s{0,20}\(/i, 'parse'],
  [/parseFloat\s{0,20}\(/i, 'parse'],
  [/Number\s{0,20}\(/i, 'transform'],
  [/String\s{0,20}\(/i, 'transform'],
  [/Boolean\s{0,20}\(/i, 'transform'],
  [/Object\.keys\s{0,20}\(/i, 'extract'],
  [/Object\.values\s{0,20}\(/i, 'extract'],
  [/Object\.entries\s{0,20}\(/i, 'extract'],
  [/Object\.assign\s{0,20}\(/i, 'merge'],
  [/\.\.\.\w+/, 'spread'],
];

const VALIDATION_OPERATION_PATTERNS: Array<[RegExp, string]> = [
  [/(validate|isValid|check|verify|assert)\s{0,20}\(/i, 'validate'],
  [/\.test\s{0,20}\(/i, 'validate'],
  [/\.match\s{0,20}\(/i, 'validate'],
  [/schema\.(validate|parse|safeParse)\s{0,20}\(/i, 'validate'],
  [/z\.\w+\s{0,20}\(/i, 'validate'],
  [/joi\.\w+/i, 'validate'],
  [/yup\.\w+/i, 'validate'],
];

const DOMAIN_PATTERNS: Array<[RegExp, string]> = [
  [/\b(user|users|account|accounts|profile|profiles|member)\b/i, 'user'],
  [/\b(auth|authentication|login|logout|signin|signout|token|session|jwt|oauth|password|credential|secret|apikey|api_key)\b/i, 'auth'],
  [/\b(payment|charge|invoice|billing|subscription|stripe|paypal)\b/i, 'payment'],
  [/\b(order|orders|cart|checkout|purchase)\b/i, 'commerce'],
  [/\b(email|mail|notification|alert|notify|message|sms)\b/i, 'notification'],
  [/\b(file|files|upload|download|attachment|blob|storage)\b/i, 'file'],
  [/\b(database|db|query|record|table|collection|document)\b/i, 'database'],
  [/\b(prisma|mongoose|sequelize|typeorm|knex)\b/i, 'database'],
  [/\b(cache|redis|memcached|cached)\b/i, 'cache'],
  [/\b(queue|job|jobs|worker|task|tasks|bull|rabbitmq)\b/i, 'queue'],
  [/\b(api|endpoint|route|routes|request|response|req|res)\b/i, 'api'],
  [/\b(webhook|webhooks|callback|hook)\b/i, 'webhook'],
  [/\b(event|events|emit|publish|subscribe|listener)\b/i, 'event'],
  [/\b(log|logs|logger|logging|trace|debug|info|warn|error)\b/i, 'logging'],
  [/\b(config|configuration|settings|options|env|environment)\b/i, 'config'],
  [/\b(test|tests|spec|describe|it\s{0,20}\(|expect\s{0,20}\()\b/i, 'test'],
];

const CODE_PATTERN_PATTERNS: Array<[RegExp, string]> = [
  [/if\s{0,20}\([^)]+\)\s{0,20}(?:return|throw)/i, 'guard_clause'],
  [/if\s{0,20}\(\s{0,20}!\s{0,20}\w+\s{0,20}\)\s{0,20}(?:return|throw)/i, 'guard_clause'],
  [/===?\s{0,20}null\b/i, 'null_check'],
  [/!==?\s{0,20}null\b/i, 'null_check'],
  [/===?\s{0,20}undefined\b/i, 'null_check'],
  [/!==?\s{0,20}undefined\b/i, 'null_check'],
  [/\?\?/i, 'null_check'],
  [/\?\s{0,20}\./i, 'null_check'],
  [/typeof\s{1,20}\w+\s{0,20}[!=]==?\s{0,20}["']undefined["']/i, 'null_check'],
  [/try\s{0,20}\{/i, 'error_handling'],
  [/catch\s{0,20}\(/i, 'error_handling'],
  [/\.catch\s{0,20}\(/i, 'error_handling'],
  [/finally\s{0,20}\{/i, 'error_handling'],
  [/throw\s{1,20}new\s{1,20}\w*Error/i, 'error_handling'],
  [/retry|retries|attempts|maxAttempts|backoff/i, 'retry_logic'],
  [/timeout|setTimeout|setInterval|clearTimeout|clearInterval/i, 'timeout'],
  [/async\s{1,20}/i, 'async_await'],
  [/await\s{1,20}/i, 'async_await'],
  [/\.then\s{0,20}\(/i, 'promise_chain'],
  [/Promise\.(all|race|allSettled|any)\s{0,20}\(/i, 'promise_composition'],
  [/new\s{1,20}Promise\s{0,20}\(/i, 'promise_creation'],
  [/cache\.(get|set|has|delete)/i, 'caching'],
  [/memoize|memo|cached/i, 'caching'],
  [/\b(page|pages|offset|limit|cursor|skip|take)\b/i, 'pagination'],
  [/\b(batch|batches|chunk|chunks)\b/i, 'batching'],
  [/\b(stream|streams|pipe|readable|writable)\b/i, 'streaming'],
  [/\b(lock|unlock|mutex|semaphore)\b/i, 'locking'],
  [/\b(rateLimit|throttle|debounce)\b/i, 'rate_limiting'],
];

const DATA_TYPE_PATTERNS: Array<[RegExp, string]> = [
  [/\[\s{0,20}\]/, 'array'],
  [/\bArray\b/, 'array'],
  [/\.length\b/, 'array'],
  [/Array\.isArray\s{0,20}\(/, 'array'],
  [/\.push\s{0,20}\(/, 'array'],
  [/\{\s{0,20}\}/, 'object'],
  [/\bObject\b/, 'object'],
  [/\.keys\s{0,20}\(/, 'object'],
  [/\.values\s{0,20}\(/, 'object'],
  [/\.entries\s{0,20}\(/, 'object'],
  [/\.hasOwnProperty\s{0,20}\(/, 'object'],
  [/['"][^'"]*['"]/, 'string'],
  [/`[^`]*`/, 'string'],
  [/\.toString\s{0,20}\(/, 'string'],
  [/\.trim\s{0,20}\(/, 'string'],
  [/\.substring\s{0,20}\(/, 'string'],
  [/\.substr\s{0,20}\(/, 'string'],
  [/\b\d+\.?\d*\b/, 'number'],
  [/Number\s{0,20}\(/, 'number'],
  [/parseInt\s{0,20}\(/, 'number'],
  [/parseFloat\s{0,20}\(/, 'number'],
  [/Math\.\w+/, 'number'],
  [/\b(true|false)\b/, 'boolean'],
  [/Boolean\s{0,20}\(/, 'boolean'],
  [/new\s{1,20}Date\s{0,20}\(/, 'date'],
  [/Date\.(now|parse)\s{0,20}\(/, 'date'],
  [/\.toISOString\s{0,20}\(/, 'date'],
  [/moment\s{0,20}\(/, 'date'],
  [/dayjs\s{0,20}\(/, 'date'],
  [/\bPromise\b/, 'promise'],
  [/\.then\s{0,20}\(/, 'promise'],
  [/async\s{1,20}/, 'promise'],
  [/await\s{1,20}/, 'promise'],
  [/\bnull\b/, 'null'],
  [/\bundefined\b/, 'undefined'],
  [/new\s{1,20}Map\s{0,20}\(/, 'map'],
  [/new\s{1,20}Set\s{0,20}\(/, 'set'],
  [/\.has\s{0,20}\(/, 'collection'],
  [/\bBuffer\b/, 'buffer'],
  [/ArrayBuffer/, 'buffer'],
  [/Uint8Array/, 'buffer'],
  [/\/[^/]+\/[gim]*/, 'regex'],
  [/new\s{1,20}RegExp\s{0,20}\(/, 'regex'],
];

// Combined operation patterns
const ALL_OPERATION_PATTERNS: Array<[RegExp, string]> = [
  ...ARRAY_OPERATION_PATTERNS,
  ...CRUD_OPERATION_PATTERNS,
  ...TRANSFORM_OPERATION_PATTERNS,
  ...VALIDATION_OPERATION_PATTERNS,
];

// ---------------------------------------------------------------------------
// SemanticAnnotator class
// ---------------------------------------------------------------------------

export class SemanticAnnotator {
  private readonly collectTiming: boolean;
  constructor(collectTiming = false) {
    this.collectTiming = collectTiming;
  }

  extractAnnotation(block: CodeBlock): SemanticAnnotation {
    const code = block.sourceCode;
    const tags = block.tags ?? [];
    let category = block.category ?? 'unknown';

    // Handle enum values (if category has a .value property)
    if (typeof category === 'object' && category !== null && 'value' in category) {
      category = (category as { value: string }).value;
    }

    const operations = this._extractOperations(code);
    const domains = this._extractDomains(code, tags);
    const patterns = this._extractPatterns(code);
    const dataTypes = this._extractDataTypes(code);
    const intent = this._inferIntent(operations, domains, patterns);

    return {
      category,
      operations,
      domains,
      patterns,
      dataTypes,
      intent,
    };
  }

  private _extractOperations(code: string): Set<string> {
    const operations = new Set<string>();
    for (const [pattern, op] of ALL_OPERATION_PATTERNS) {
      if (pattern.test(code)) operations.add(op);
    }
    return operations;
  }

  private _extractDomains(code: string, tags: string[]): Set<string> {
    const domains = new Set<string>();
    const text = code + ' ' + tags.join(' ');
    for (const [pattern, domain] of DOMAIN_PATTERNS) {
      if (pattern.test(text)) domains.add(domain);
    }
    return domains;
  }

  private _extractPatterns(code: string): Set<string> {
    const patterns = new Set<string>();
    for (const [pattern, name] of CODE_PATTERN_PATTERNS) {
      if (pattern.test(code)) patterns.add(name);
    }
    return patterns;
  }

  private _extractDataTypes(code: string): Set<string> {
    const dataTypes = new Set<string>();
    for (const [pattern, dtype] of DATA_TYPE_PATTERNS) {
      if (pattern.test(code)) dataTypes.add(dtype);
    }
    return dataTypes;
  }

  private _inferIntent(
    operations: Set<string>,
    domains: Set<string>,
    patterns: Set<string>
  ): string {
    const parts: string[] = [];

    if (operations.size > 0) {
      parts.push([...operations].sort().join('+'));
    }
    if (domains.size > 0) {
      parts.push('on:' + [...domains].sort().join('+'));
    }
    if (patterns.size > 0) {
      parts.push('with:' + [...patterns].sort().join('+'));
    }

    return parts.length > 0 ? parts.join('|') : 'unknown';
  }
}

/** Convenience: create annotation for a single block */
export function extractAnnotation(block: CodeBlock): SemanticAnnotation {
  return new SemanticAnnotator().extractAnnotation(block);
}

/** Convert SemanticAnnotation sets to sorted arrays for JSON serialization */
export function annotationToDict(ann: SemanticAnnotation): Record<string, unknown> {
  return {
    category: ann.category,
    operations: [...ann.operations].sort(),
    domains: [...ann.domains].sort(),
    patterns: [...ann.patterns].sort(),
    data_types: [...ann.dataTypes].sort(),
    intent: ann.intent,
  };
}
